### PR TITLE
transports/quic: Refactor listener handling (updated)

### PR DESCRIPTION
**Pull request by elenaf9**
# Description

_Replaces #17. Rebased onto latest `kpp::libp2p-quic` and merged latest `libp2p::main`. Apart form that, the only other change compared to #17 is to remove the `Mutex` in `InAddr` (ef7b823)._

_Copying over the description from #17:_
 
Discussed in https://github.com/libp2p/rust-libp2p/pull/2289#discussion_r907228272.
This PR adapts the quic transport to the upstream listener changes of libp2p/rust-libp2p#2652. Additionally, it now allows the quic transport to listen on multiple addresses / endpoints. The changes include:
- `QuicTransport` now owns and drives multiple endpoints. An endpoint is created each time `Transport::listen_on` is called.
- When dialing a remote, a random existing listening endpoint is used. If none exists a new one is created that only supports outbound connections (i.e. server config is `None`). If then a new listening endpoint is created any following dials will use the new endpoint, but existing connections are not migrated.

## Open Questions

1. If a first listening endpoint is created and we already have an endpoint used for dialing (without server capabilities enabled): should we, instead of creating a new endpoint, enable server capabilities on the old endpoint and rebind it to the new listening address? The benefit of this would be that the remote peers could then update their address information once they notice the address change. For later dial attempts they would then know the correct address to dial.

Quoting response from @mxinden:
> This sounds very neat to me, especially the part where the remote is notified of the address change. I am fine with this landing in a later iteration in case it is a big undertaking.

-> I will look into it and come back to this.

---

2. Rare case: if the user starts listening on a socket address that we already randomly chose for an existing dialer-endpoint, what should happen?

Quoting again response from @mxinden:
> I would be fine with not handling this case, that is erroring on the listen_on call, forcing the user to handle the case.
>
> I am guessing re-using the existing dialer-endpoint is complicated, especially when handling listen requests across interfaces (i.e. 0.0.0.0) as we would need to do an additional listen_on call for the remaining interfaces with the specified port, rig

-> If we decide to add the logic mentioned in question 1. to migrate a dialing endpoint to be the new listener, we have to deal the case of one interface vs multiple anyway. So it probably makes sense to solve this together. Will look into it.

### CODE REVIEW

File: transports/quic/src/endpoint.rs
Comment (elenaf9) [inline transports/quic/src/endpoint.rs]: @mxinden:
> Just double checking here, does the use of IPv4Addr::LOCALHOST allow to dial endpoints not colocated on the same host?

File: transports/quic/src/transport.rs
Comment (elenaf9) [inline transports/quic/src/transport.rs]: @mxinden:
> I guess you could re-use the logic used in libp2p-tcp?
>
> https://github.com/libp2p/rust-libp2p/blob/2f2b7cbec127a48d869d87ad52d69a4c4c29ddb5/transports/tcp/src/lib.rs#L128-L158

File: transports/quic/src/transport.rs
Comment (elenaf9) [inline transports/quic/src/transport.rs]: @kpp:
> Are we okay with an infinite recursion in case there are infinite many IPv6 events when we are bound to IPv4?

Comment (elenaf9) [reply transports/quic/src/transport.rs]: That can only happen if the `IfWatcher` will always return `Poll::Ready(ipv6_event)` if we poll it, which would be a problematic bug either way, no?

I'd like to catch the case here that the `IfWatcher` first reports an IPv6 address and directly afterwards an Ipv4 one. In that situation we can not just return `Poll::Pending` here but instead have to solve it by either a) wrapping the whole block in a loop and `continue` if an Ipv6 address is reported, or b) do this recursion. In both cases the behaviour that you described would be a problem. No sure if this can be solved any other way?

File: transports/quic/src/endpoint.rs
Comment (elenaf9) [inline transports/quic/src/endpoint.rs:139]: @kpp:
> I believe this one can be polled directly, right?

**Review by elenaf9 — COMMENTED** [review]
Copied over the open discussions from #17. Lmk if I forgot something.

File: transports/quic/src/endpoint.rs
Comment (elenaf9) [reply transports/quic/src/endpoint.rs:139]: If I understand you correctly, you are proposing to have the logic of the `background_task` directly in `Listener::poll_next` instead of a separate task, right?

I considered doing this, but after an out of band discussion with @mxinden decided against it. 
The reason for this is that every time that any event / reading / writing / udp packet on any connection happens, the endpoint is involved and thus the waker of its task will be called. `Listener::poll` is driven by `Transport::poll`, which in return is polled directly in the `Swarm`'s main loop that also drives the behaviours and the `Pool`. Thus they all share the same waker. Right now if one of these structures triggers the waker, all of them are polled again. If we drive the endpoint is this main loop, this would create unnecessary overhead. 
Does that explanation make sense?  
I agree that this is not ideal; alternative ideas would be appreciated.

**Comment by mxinden** [comment]
@elenaf9 thanks for porting all the past discussions over here. As far as I can tell there is no input needed from my side right now. Please correct me if I am wrong. Also please let me know in case you want me to review any of these changes again.

**Comment by kpp** [comment]
I tested it:
libp2p-perf: works :+1:
substrate: works but same issue: server is not notified when client drops

Also one minor issue: I got `Error while sending on QUIC UDP socket: Os { code: 97, kind: Uncategorized, message: "Address family not supported by protocol" }` when sending data from `0.0.0.0` to `::1`.

LGTM. Other issues can be solved in follow up PRs.

File: transports/quic/src/transport.rs
Comment (kpp) [reply transports/quic/src/transport.rs]: > which would be a problematic bug either way, no?

No. You can delete ipv4 from an interface, so you will always get ipv6 events from IfWatcher. An infinite loop will crash the code, a missing ipv4 addr will not.

**Comment by mxinden** [comment]
> Also one minor issue: I got `Error while sending on QUIC UDP socket: Os { code: 97, kind: Uncategorized, message: "Address family not supported by protocol" }` when sending data from `0.0.0.0` to `::1`.

Do I understand correctly that you only had the `0.0.0.0` listen address, or also `::/0` @kpp?

File: transports/quic/src/transport.rs
Comment (mxinden) [reply transports/quic/src/transport.rs]: > > which would be a problematic bug either way, no?
> 
> No. You can delete ipv4 from an interface, so you will always get ipv6 events from IfWatcher. An infinite loop will crash the code, a missing ipv4 addr will not.

I am not sure I am following @kpp. Are you saying that if-watch could potentially return ipv6 events non-stop, i.e. without a single `Poll::Pending` in between?

> a) wrapping the whole block in a loop and `continue` if an Ipv6 address is reported,

Note that Rust does not have tail recursion, thus large recursive nesting could potentially be problematic in regards to the stack size. I doubt this would be a problem here, in other words I expect this to only recurse little.

From a different perspective, I don't think recursion is idiomatic Rust. Thus I would favor a `loop`. What do you think @elenaf9?

See also https://github.com/libp2p/rust-libp2p/pull/2638#discussion_r873945621

Comment (kpp) [reply transports/quic/src/transport.rs]: > without a single Poll::Pending in between?

You are right =) thanks

**Comment by kpp** [comment]
I had two nodes, one is `0.0.0.0` only, the other is `::1` only.

File: transports/quic/src/endpoint.rs
Comment (kpp) [reply transports/quic/src/endpoint.rs]: > does the use of IPv4Addr::LOCALHOST allow to dial

No. In this case we get error:

```
ERROR libp2p_quic::endpoint: Error while sending on QUIC UDP socket: Os { code: 22, kind: InvalidInput, message: "Invalid argument" }
```

Comment (elenaf9) [reply transports/quic/src/endpoint.rs]: Thanks for looking into this @kpp! Would you mind testing it with `Ipv4Addr::UNSPECIFIED`, which I assume would fix this?

Comment (kpp) [reply transports/quic/src/endpoint.rs]: I did and it worked

File: transports/quic/src/transport.rs
Comment (elenaf9) [reply transports/quic/src/transport.rs]: > From a different perspective, I don't think recursion is idiomatic Rust. Thus I would favor a loop. What do you think @elenaf9?

Done by @kpp in elenaf9#5 / 574d534.

File: transports/quic/src/endpoint.rs
Comment (elenaf9) [reply transports/quic/src/endpoint.rs]: Fixed in 1a4b8ac.

File: transports/quic/src/transport.rs
Comment (elenaf9) [reply transports/quic/src/transport.rs]: 👍 Done in 1a4b8ac.

**Comment by elenaf9** [comment]
Thank you for the reviews @mxinden @kpp!

> Also one minor issue: I got Error while sending on QUIC UDP socket: Os { code: 97, kind: Uncategorized, message: "Address family not supported by protocol" } when sending data from 0.0.0.0 to ::1.

Fixed the logic now for selecting an endpoint for dialing:
Analogous to how socket reuse is done in the tcp transport, we now look in our existing listeners for an endpoint with matching socket family (Ipv4 / Ipv6) and loopback interface (`is_loopback`). If none is found we create a new dialer that is listening on a wildcard address with the matching socket family. Thus we now have two dialing endpoints, one for Ipv4 and one for Ipv6 addresses. 
@kpp thanks for adding the `ipv4_dial_ipv6` test! It now passes. 

---
From my side this is ready for merge. I can not merge the latest upstream master because of some more muxer changes. @kpp I can look into fixing them, unless you are already working on it?

Edit: CI is failing because of libp2p#2770 which is independent of libp2p-quic, but came up because we use libp2p-core as dependency.

**Comment by kpp** [comment]
> I can look into fixing them, unless you are already working on it?

Sure. I tried but it was a little bit hard because your branch was not merged yet.

You can take this issue while I am working on that issue with disconnects.

**Comment by elenaf9** [comment]
Sorry for the force-push. I did not change anything, just reverted the merge of upstream master. Will do a separate PR that includes the latest master and adapts the QuicMuxer.

---

Referencing a comment in https://github.com/libp2p/rust-libp2p/issues/2722:
@kpp 
> I believe I should try to move QuicMuxer to a new API to give it a try.

To avoid duplicating work: are you working on this right now? From your above comment (https://github.com/kpp/rust-libp2p/pull/19#issuecomment-1193353839) I assumed that you are currently working on the disconnect issues, but I'm happy to sync on this in case that you have now started looking into this. Otherwise I'd draft the PR and we can discuss the solution there.

**Comment by kpp** [comment]
> To avoid duplicating work: are you working on this right now?

No.

**Comment by elenaf9** [comment]
> Sorry for the force-push. I did not change anything, just reverted the merge of upstream master. Will do a separate PR that includes the latest master and adapts the QuicMuxer.
> 
> Referencing a comment in [libp2p#2722](https://github.com/libp2p/rust-libp2p/issues/2722): @kpp
> 
> > I believe I should try to move QuicMuxer to a new API to give it a try.
> 
> To avoid duplicating work: are you working on this right now? From your above comment ([#19 (comment)](https://github.com/kpp/rust-libp2p/pull/19#issuecomment-1193353839)) I assumed that you are currently working on the disconnect issues, but I'm happy to sync on this in case that you have now started looking into this. Otherwise I'd draft the PR and we can discuss the solution there.

See elenaf9#6.

**Comment by elenaf9** [comment]
For the sake of reducing the number of parallel open PRs related to quic, and also to ease reviewing: 
I think we can already review and merge this PR independently of elenaf9#6, since they target different parts of the implementation. Wdyt @kpp? Even though it does not include all of the latest master changes, I think it is still an improvement.

Additional reviews  would also be appreciated (cc @mxinden @thomaseizinger @marcopolo;  relevent is only the diff within transports/quic).

File: transports/quic/src/endpoint.rs
Comment (mxinden) [inline transports/quic/src/endpoint.rs:61]: :rocket: great to no longer having to pass a `Multiaddr`.

**Review by mxinden — APPROVED** [review]
I am in favor of merging here as is. Thanks for the ping and thanks for staying on top of things @elenaf9.

**Comment by kpp** [comment]
I think we should merge it. I will review it today/tomorrow, wait for an additional reviews till the end of Monday and merge it.

**Comment by kpp** [comment]
<details> 
  <summary>bench `multiple-endpoints-2#bf06d967`</summary>

```
# Start Rust and Golang servers.
Local peer id: PeerId("Qmcqq9TFaYbb94uwdER1BXyGfCFY4Bb1gKozxNyVvLvTSw")
about to listen on "/ip4/127.0.0.1/udp/9992/quic"
Listening on "/ip4/127.0.0.1/udp/9992/quic".

# Rust -> Rust

Local peer id: PeerId("12D3KooWAmxKDBCLbYU8juAmRhUKGhEYcg79c4hCnvENraAosWkD")
IncomingConnection { local_addr: "/ip4/127.0.0.1/udp/9992/quic", send_back_addr: "/ip4/127.0.0.1/udp/40500/quic" }
ConnectionEstablished { peer_id: PeerId("12D3KooWAmxKDBCLbYU8juAmRhUKGhEYcg79c4hCnvENraAosWkD"), endpoint: Listener { local_addr: "/ip4/127.0.0.1/udp/9992/quic", send_back_addr: "/ip4/127.0.0.1/udp/40500/quic" }, num_established: 1, concurrent_dial_errors: None }
Behaviour(PerfRunDone(10.000694835s, 2232061300))
Interval	Transfer	Bandwidth
0 s - 10.00 s	2232 MBytes	1785.29 MBit/s
ConnectionClosed { peer_id: PeerId("Qmcqq9TFaYbb94uwdER1BXyGfCFY4Bb1gKozxNyVvLvTSw"), endpoint: Dialer { address: "/ip4/127.0.0.1/udp/9992/quic", role_override: Dialer }, num_established: 0, cause: None }

# Rust -> Golang

Local peer id: PeerId("12D3KooWPY3YCErc99qJ74h363qjHHf93DFuuK8U5n2SPGJHooxD")
Interval	Transfer	Bandwidth
0 s - 10.00 s	2056 MBytes	1644.61 MBit/s
ConnectionClosed { peer_id: PeerId("12D3KooWL3XJ9EMCyZvmmGXL2LMiVBtrVa2BuESsJiXkSj7333Jw"), endpoint: Dialer { address: "/ip4/127.0.0.1/udp/9993/quic", role_override: Dialer }, num_established: 0, cause: None }


# Golang -> Rust

2022/08/22 15:11:51 failed to sufficiently increase receive buffer size (was: 208 kiB, wanted: 2048 kiB, got: 416 kiB). See https://github.com/lucas-clemente/quic-go/wiki/UDP-Receive-Buffer-Size for details.
IncomingConnection { local_addr: "/ip4/127.0.0.1/udp/9992/quic", send_back_addr: "/ip4/127.0.0.1/udp/32942/quic" }
ConnectionEstablished { peer_id: PeerId("12D3KooWEiWwtDiNQcwynr4TUzY3m1zFVJ9YDVifDb7FQBajCt9Y"), endpoint: Listener { local_addr: "/ip4/127.0.0.1/udp/9992/quic", send_back_addr: "/ip4/127.0.0.1/udp/32942/quic" }, num_established: 1, concurrent_dial_errors: None }
Interval 	Transfer	Bandwidth
0s - 10.00 s 	1672 MBytes	1337.58 MBit/s
Behaviour(PerfRunDone(9.999930943s, 1672064000))

# Golang -> Golang

2022/08/22 15:12:01 failed to sufficiently increase receive buffer size (was: 208 kiB, wanted: 2048 kiB, got: 416 kiB). See https://github.com/lucas-clemente/quic-go/wiki/UDP-Receive-Buffer-Size for details.
Interval 	Transfer	Bandwidth
0s - 10.00 s 	1681 MBytes	1344.74 MBit/s

```

</details>

**Comment by thomaseizinger** [comment]
> Additional reviews would also be appreciated (cc @mxinden @thomaseizinger @MarcoPolo; relevent is only the diff within transports/quic).

I am not gonna be able to fit a review in in time but I think this has had a good few pairs of eyes. Happy to see it merged. The numbers look good too :)

**Comment by mxinden** [comment]
Friendly ping @kpp. Would you mind hitting the merge button?

**Comment by kpp** [comment]
@mxinden I am not quite sure whether we have to merge it right now because of https://github.com/elenaf9/rust-libp2p/pull/6#issuecomment-1222292596.

**Comment by elenaf9** [comment]
> @mxinden I am not quite sure whether we have to merge it right now because of [elenaf9#6 (comment)](https://github.com/elenaf9/rust-libp2p/pull/6#issuecomment-1222292596).

I think we can already merge this PR (if you have reviewed and approved) because it does not depend on elenaf9#6. The issue about not being compatible with the go-impl  that appears on that other PR is not an issue here. 
We don't per-se _need_ to already merge it, but reducing the overall number of pending PRs related to QUIC makes it easier for folks on the outside to understand what is happening where. Any reason why you would prefer to wait @kpp?

**Pull request merged**
